lib/liquid/expression.rb

```diff
@@ -45,4 +47,109 @@ def self.parse(markup)
       end
     end
   end
+
+  class Expression2
+    LITERALS = {
+      nil => nil,
+      'nil' => nil,
+      'null' => nil,
+      '' => nil,
+      'true' => true,
+      'false' => false,
+      'blank' => '',
+      'empty' => '',
+      '-' => VariableLookup.parse("-")
+    }.freeze
+
+    DOT = ".".ord
+    ZERO = "0".ord
+    NINE = "9".ord
+    DASH = "-".ord
+
+    # Use an atomic group (?>...) to avoid pathological backtracing from
+    # malicious input as described in https://github.com/Shopify/liquid/issues/1357
+    RANGES_REGEX = /\A\(\s*(?>(\S+)\s*\.\.)\s*(\S+)\s*\)\z/
+    CACHE = LruRedux::Cache.new(10_000) # most themes would have less than 2,000 unique expression
+
+    class << self
+      def string_scanner
+        @ss ||= StringScanner.new("")
+      end
+
+      def parse(markup)
+        return unless markup
+
+        markup = markup.strip # markup can be a frozen string
+
+        if (markup.start_with?('"') && markup.end_with?('"')) ||
+          (markup.start_with?("'") && markup.end_with?("'"))
+          return markup[1..-2]
+        elsif LITERALS.key?(markup)
+          return LITERALS[markup]
+        end
+
+        return CACHE[markup] if CACHE.key?(markup)
+
+        CACHE[markup] = inner_parse(markup)
+      end
+
+      def inner_parse(markup)
+        if (markup.start_with?("(") && markup.end_with?(")")) && markup =~ RANGES_REGEX
+          return RangeLookup.parse(Regexp.last_match(1), Regexp.last_match(2))
+        end
+
+        if (num = parse_number(markup))
+          num
+        else
+          VariableLookup.parse(markup)
+        end
+      end
+
+      def parse_number(markup)
+        ss = string_scanner
+        ss.string = markup
```

Comment:
Great catch!
I agree that `StringScanner` doesn't fit for this use case 👍 

Unfortunately, I didn't get to see any performance improvement as much as I hoped to see... (roughly 2% diff by keeping the cursor in Ruby ☹️)

(10 seconds warmup / 20 seconds run)
**Baseline**
```
ruby 3.4.0dev (2024-04-23T16:59:11Z v3.4.0-pshopify-pr.. 70b931820e) +YJIT [arm64-darwin23]
Warming up --------------------------------------
Liquid::Expression2#parse: number
                        39.927k i/100ms
Calculating -------------------------------------
Liquid::Expression2#parse: number
                        405.480k (± 0.4%) i/s    (2.47 μs/i) -      8.145M in  20.087950s
```

**Without `StringScanner`**
```
ruby 3.4.0dev (2024-04-23T16:59:11Z v3.4.0-pshopify-pr.. 70b931820e) +YJIT [arm64-darwin23]
Warming up --------------------------------------
Liquid::Expression2#parse: number
                        39.051k i/100ms
Calculating -------------------------------------
Liquid::Expression2#parse: number
                        395.787k (± 0.3%) i/s    (2.53 μs/i) -      7.927M in  20.029537s
```

<details>
<summary>Source code</summary>

```ruby
def parse_number(markup)
  # the first byte must be a digit, a period, or  a dash
  byte = markup.getbyte(0)
  return false if byte != DASH && byte != DOT && (byte < ZERO || byte > NINE)

  is_integer = true
  last_dot_pos = nil
  num_end_pos = nil
  idx = 1

  while byte = markup.getbyte(idx)
    return false if byte != DOT && (byte < ZERO || byte > NINE)

    idx += 1

    # we found our number and now we are just scanning the rest of the string
    next if num_end_pos

    if byte == DOT
      if is_integer == false
        num_end_pos = idx - 1
      else
        is_integer = false
        last_dot_pos = idx
      end
    end
  end

  num_end_pos = markup.length if idx == markup.length

  return markup.to_i if is_integer

  if num_end_pos
    # number ends with a number "123.123"
    markup.byteslice(0, num_end_pos).to_f
  elsif last_dot_pos
    markup.byteslice(0, last_dot_pos).to_f
  else
    # we should never reach this point
    false
  end
end
```
</details>

Reply:
We need to profile to find what's slow, but I think avoiding the mutation of `StringScanner` is a great improvement.  Maybe try profiling with Vernier?

If I'm reading this correctly, it looks like this function requires that the whole string must contain `0-9` or a `.` character.  We seem to exit early if `return false if byte != DOT && (byte < ZERO || byte > NINE)` happens.

If that's true, then we could do a fast path case that's like `return markup.to_i if markup.match?(/\A[0-9]+\z/)`.

Something like this:

<details>

```ruby
def parse_number(markup)
  # the first byte must be a digit, a period, or  a dash
  byte = markup.getbyte(0)
  return false if byte != DASH && byte != DOT && (byte < ZERO || byte > NINE)

  # This handles all integer cases
  return markup.to_i if markup.match?(/\A[0-9]+\z/)

  # rest of method
end
```

</details>

It might be possible to do the float parsing without the loop, but I don't really understand what it's doing.  It looks like it's trying to handle a number with multiple dots in it?  I don't understand the case.

Reply:
> It looks like it's trying to handle a number with multiple dots in it? I don't understand the case.
Yup, since `"123.456.789".to_f` returns `123.456` in Ruby, we need to replicate that logic in Liquid.

The original `Tokenizer` actually parses the number with the Regexs:
```ruby
...
case markup
when INTEGERS_REGEX
  Regexp.last_match(1).to_i
when RANGES_REGEX
  RangeLookup.parse(Regexp.last_match(1), Regexp.last_match(2))
when FLOATS_REGEX
  Regexp.last_match(1).to_f
else
...
```
with this approach of using Regex, we are scanning the String twice for parsing a number. (Once with `INTEGERS_REGEX` and another one with `FLOATS_REGEX`)
By scanning the String once, `Tokenizer2` is 1.7x faster than `Tokenizer1` for parsing numbers.

Reply:
> Yup, since `"123.456.789".to_f` returns `123.456` in Ruby, we need to replicate that logic in Liquid.

Do we really need to replicate all logic from Ruby?  For example `"1e10".to_f` works in Ruby, but it's not supported here.  Also, if it's following the same rules [that the lexer uses](https://github.com/Shopify/liquid/blob/200fa0c11baebf110624ed0f24a7136d7e2f2870/lib/liquid/lexer.rb#L22), I don't think we would encounter a `:number` token that contains multiple `.`:

```
$ ruby -I lib -r liquid/lexer -e'p Liquid::Lexer.new("12.34.56").tokenize'
[[:number, "12.34"], [:dot, "."], [:number, "56"], [:end_of_string]]
```

Are these strings coming from the lexer or somewhere else?